counsel.el

```diff
@@ -4918,6 +4918,36 @@ An extra action allows to switch to the process buffer."
 ;; `counsel-slime-repl-history' within
 ;; `counsel--browse-history-action'.
 
+;;** `counsel-esh-dir-history'
+(declare-function eshell/cd "eshell")
+
+(defun counsel--esh-dir-history-action-cd (pair)
+  "Default action for counsel-esh-dir-history. It changes the
+current working directory to the one selected by the user."
+  (eshell/cd (car pair)))
+
+(defun counsel--esh-dir-history-action-edit (pair)
+  "Action for counsel-esh-dir-history to insert the selected
+directory to the Eshell buffer prefixed by \"cd \", allowing the
+caller to modify parts of the directory before switching to it."
+  (ivy--action-insert (format "cd %s" (car pair))))
+
+(defvar eshell-last-dir-ring)
```

Comment:
Better to move this into the body of `counsel-esh-dir-history`, to localise its effect.

Reply:
Sure, it has been moved.